nucypher/acumen/perception.py

```diff
@@ -30,146 +32,320 @@
+        return self._current_state.nickname
 
     @property
     def icon(self) -> str:
-        if self.nickname is NO_KNOWN_NODES:
-            return str(NO_KNOWN_NODES)
-        return self.nickname.icon
+        # FIXME: should it be called at all if there are no states recorded?
+        return self._current_state.icon
 
-    def addresses(self):
-        return self._nodes.keys()
+    @property
+    def timestamp(self):
+        return self._current_state.timestamp
 
-    def snapshot(self):
-        fleet_state_checksum_bytes = binascii.unhexlify(self.checksum)
-        fleet_state_updated_bytes = self.updated.epoch.to_bytes(4, byteorder="big")
-        return fleet_state_checksum_bytes + fleet_state_updated_bytes
-
-    def record_fleet_state(self, additional_nodes_to_track=None):
-        if additional_nodes_to_track:
-            self.additional_nodes_to_track.extend(additional_nodes_to_track)
-
-        if not self._nodes:
-            # No news here.
-            return
-        sorted_nodes = self.sorted()
-
-        sorted_nodes_joined = b"".join(bytes(n) for n in sorted_nodes)
-        checksum = keccak_digest(sorted_nodes_joined).hex()
-        if checksum not in self.states:
-            self.checksum = keccak_digest(b"".join(bytes(n) for n in self.sorted())).hex()
-            self.updated = maya.now()
-            # For now we store the sorted node list.  Someday we probably spin this out into
-            # its own class, FleetState, and use it as the basis for partial updates.
-            new_state = self.FleetState(nickname=self.nickname,
-                                        nodes=sorted_nodes,
-                                        icon=self.icon,
-                                        updated=self.updated,
-                                        checksum=self.checksum)
-            self.states[checksum] = new_state
-            return checksum, new_state
-
-    def start_tracking_state(self, additional_nodes_to_track=None):
-        if additional_nodes_to_track is None:
-            additional_nodes_to_track = list()
-        self.additional_nodes_to_track.extend(additional_nodes_to_track)
-        self._tracking = True
-        self.update_fleet_state()
-
-    def sorted(self):
-        nodes_to_consider = list(self._nodes.values()) + self.additional_nodes_to_track
-        return sorted(nodes_to_consider, key=lambda n: n.checksum_address)
+    def items(self):
+        return self._current_state.items()
 
-    def shuffled(self):
-        nodes_we_know_about = list(self._nodes.values())
-        random.shuffle(nodes_we_know_about)
-        return nodes_we_know_about
+    def values(self):
+        return self._current_state.values()
+
+    def latest_states(self, quantity):
+        """
+        Returns at most ``quantity`` latest archived states (including the current one),
+        in chronological order.
+        """
+        return self._archived_states[-min(len(self._archived_states), quantity):]
```

Comment:
Consider using a named variable here so that I have a place to trigger my debugger 😄

Reply:
Fixed, but can't you set a breakpoint on a `return`?